fairmq/shmem/Common.h

```diff
@@ -41,13 +44,126 @@ using RBTreeBestFitSegment = boost::interprocess::basic_managed_shared_memory<ch
     boost::interprocess::null_index>;
     // boost::interprocess::iset_index>;
 
+inline std::string MakeShmName(const std::string& shmId, const std::string& type) {
+    return std::string("fmq_" + shmId + "_" + type);
+}
+
+inline std::string MakeShmName(const std::string& shmId, const std::string& type, int index) {
+    return std::string(MakeShmName(shmId, type) + "_" + std::to_string(index));
+}
+
+struct RefCount
+{
+    explicit RefCount(uint16_t c)
+        : count(c)
+    {}
+
+    uint16_t Get() { return count.load(); }
+    uint16_t Increment() { return count.fetch_add(1); }
+    uint16_t Decrement() { return count.fetch_sub(1); }
+
+    std::atomic<uint16_t> count;
+};
+
+// Number of nodes allocated at once when the allocator runs out of nodes.
+static constexpr size_t numNodesPerBlock = 4096;
+// Maximum number of totally free blocks that the adaptive node pool will hold.
+// The rest of the totally free blocks will be deallocated with the segment manager.
+static constexpr size_t maxFreeBlocks = 2;
+
+using RefCountPool = boost::interprocess::adaptive_pool<RefCount, boost::interprocess::managed_shared_memory::segment_manager, numNodesPerBlock, maxFreeBlocks>;
+
 using SegmentManager = boost::interprocess::managed_shared_memory::segment_manager;
 using VoidAlloc      = boost::interprocess::allocator<void, SegmentManager>;
 using CharAlloc      = boost::interprocess::allocator<char, SegmentManager>;
 using Str            = boost::interprocess::basic_string<char, std::char_traits<char>, CharAlloc>;
 using StrAlloc       = boost::interprocess::allocator<Str, SegmentManager>;
 using StrVector      = boost::interprocess::vector<Str, StrAlloc>;
 
+// ShmHeader stores user buffer alignment and the reference count in the following structure:
+// [HdrOffset(uint16_t)][Hdr alignment][Hdr][user buffer alignment][user buffer]
+// The alignment of Hdr depends on the alignment of std::atomic and is stored in the first entry
+struct ShmHeader
+{
+    struct Hdr
+    {
+        uint16_t userOffset;
+        std::atomic<uint16_t> refCount;
+    };
+
+    static Hdr* HdrPtr(char* ptr)
+    {
```

Comment:
```suggestion
    static Hdr* HdrPtr(char* ptr)
    {
        assert(ptr);
```

Basically for all of the member functions that take raw pointer args and assume they are non-null, right?

Reply:
I don't wanna touch this part in this PR. It was just move from one place to another.